### PR TITLE
Add docs for Dockerfile ADD/COPY --chown flag

### PR DESCRIPTION
**Pull request by estesp**
Document the new `--chown` flag added to the ADD and COPY commands in
the Dockerfile format.

Signed-off-by: Phil Estes <estesp@linux.vnet.ibm.com>

The actual implementation is ready to merge in moby/moby#34263

### CODE REVIEW

File: docs/reference/builder.md [hunk 1]
```diff
@@ -817,14 +817,14 @@ change them using `docker run --env <key>=<value>`.
...
+- `ADD [--chown <user>:<group>] <src>... <dest>`
+- `ADD [--chown <user>:<group>] ["<src>",... "<dest>"]` (this form is required for paths containing
 whitespace)
 
 The `ADD` instruction copies new files, directories or remote file URLs from `<src>`
 and adds them to the filesystem of the image at the path `<dest>`.
 
-Multiple `<src>` resource may be specified but if they are files or
+Multiple `<src>` resources may be specified but if they are files or
 directories then they must be relative to the source directory that is
```
Comment (mstanleyjones) [inline docs/reference/builder.md]: directories, they must be

Also, "relative" I think means that they must be subdirectories of the build context? Maybe you can reword this to make it clearer.

File: docs/reference/builder.md [hunk 2]
```diff
@@ -848,7 +848,25 @@ named `arr[0].txt`, use the following;
     ADD arr[[]0].txt /mydir/    # copy a file named "arr[0].txt" to /mydir/
 
 
-All new files and directories are created with a UID and GID of 0.
+All new files and directories are created with a UID and GID of 0, unless the
+optional `--chown` flag is used to provide a user and group definition which
+will be used to change the ownership of the created files and directories. The
```
Comment (mstanleyjones) [inline docs/reference/builder.md]: How about "optional `--chown` flag specifies a given username, groupname, or UID/GID combination."

(if you only want to give a GID, how does Docker know it's not a UID? I think if you only provide a single integer it's always assumed to be a UID. Maybe best to state this too.)

**Comment by mstanleyjones** [comment]
cc/ @gbarr01 for FYI

File: docs/reference/builder.md [hunk 3]
```diff
@@ -817,14 +817,14 @@ change them using `docker run --env <key>=<value>`.
 
 ADD has two forms:
 
-- `ADD <src>... <dest>`
-- `ADD ["<src>",... "<dest>"]` (this form is required for paths containing
+- `ADD [--chown <user>:<group>] <src>... <dest>`
```
Comment (shouze) [inline docs/reference/builder.md]: Sure about that syntax? Not working for me & @presto9292 see https://github.com/moby/moby/pull/34263#issuecomment-328034492

Comment (estesp) [reply docs/reference/builder.md]: Thanks @shouze ; you are correct that it requires an `=`; I will update this PR as it also has other edits required before being complete.

**Comment by estesp** [comment]
@mstanleyjones thanks for the review; I have updated per your comments; PTAL!

@shouze examples and flag documentation updated with the *correct* argument style--thanks for the catch!

**Comment by rcjsuen** [comment]
What is the behaviour of using this flag on Windows as the `/etc/passwd` and `/etc/group` files do not exist there? Should this be included in the documentation?

File: docs/reference/builder.md [hunk 4]
```diff
@@ -848,7 +848,26 @@ named `arr[0].txt`, use the following;
     ADD arr[[]0].txt /mydir/    # copy a file named "arr[0].txt" to /mydir/
 
 
-All new files and directories are created with a UID and GID of 0.
+All new files and directories are created with a UID and GID of 0, unless the
```
Comment (mstanleyjones) [inline docs/reference/builder.md]: How do the concepts of UID and GID 0 map to Windows?

File: docs/reference/builder.md [hunk 5]
```diff
@@ -848,7 +848,26 @@ named `arr[0].txt`, use the following;
...
+optional `--chown` flag specifies a given username, groupname, or UID/GID
+combination to request specific ownership of the content added. The
+format of the `--chown` flag allows for either username and groupname strings
+or direct integer UID and GID in any combination. Providing a username without
+groupname or a UID without GID will use the name numeric UID as the GID. If a
+username or groupname is provided, the container's root filesystem
+`/etc/passwd` and `/etc/group` files will be used to perform the translation
```
Comment (mstanleyjones) [inline docs/reference/builder.md]: How does this map to Windows?

File: docs/reference/builder.md (hunk 4 above)
Comment (estesp) [reply docs/reference/builder.md]: They do not apply/map at all.. I assume we need a note about that :)  A `Dockerfile` with the `--chown` flag on an `ADD` or `COPY` command will fail during `docker build` if it is built on a Windows Docker daemon.

This was discussed in the PR for the feature implementation itself--the other option was to "ignore" the flag on Windows but today this is **not** the way it is implemented.

File: docs/reference/builder.md [hunk 6]
```diff
@@ -848,7 +848,26 @@ named `arr[0].txt`, use the following;
     ADD arr[[]0].txt /mydir/    # copy a file named "arr[0].txt" to /mydir/
 
 
-All new files and directories are created with a UID and GID of 0.
+All new files and directories are created with a UID and GID of 0, unless the
+optional `--chown` flag specifies a given username, groupname, or UID/GID
+combination to request specific ownership of the content added. The
+format of the `--chown` flag allows for either username and groupname strings
+or direct integer UID and GID in any combination. Providing a username without
+groupname or a UID without GID will use the name numeric UID as the GID. If a
```
Comment (thaJeztah) [inline docs/reference/builder.md]: s/name/same/ (probably)

File: docs/reference/builder.md (hunk 4 above)
Comment (thaJeztah) [reply docs/reference/builder.md]: Perhaps we should add a header above this section; both for easier linking (anchor), and an introduction "This is a Linux-only feature". Also, the sections for `COPY` and `ADD` for this feature could then be combined if we think that's a good idea (_"All new files and directories that are added through `ADD` or `COPY` ...."_).  (I'm sure @mstanleyjones has better suggestions :smile:)

What is actually the owner of the files on Windows? (@johnstep?)

File: docs/reference/builder.md [hunk 7]
```diff
@@ -968,7 +988,26 @@ named `arr[0].txt`, use the following;
 
     COPY arr[[]0].txt /mydir/    # copy a file named "arr[0].txt" to /mydir/
 
-All new files and directories are created with a UID and GID of 0.
+All new files and directories are created with a UID and GID of 0, unless the
+optional `--chown` flag specifies a given username, groupname, or UID/GID
+combination to request specific ownership of the copied content. The
+format of the `--chown` flag allows for either username and groupname strings
+or direct integer UID and GID in any combination. Providing a username without
+groupname or a UID without GID will use the name numeric UID as the GID. If a
```
Comment (thaJeztah) [inline docs/reference/builder.md]: s/name/same/

**Comment by mstanleyjones** [comment]
I like the idea of a note like "The `--chown` and `--chgrp` features are only supported on Dockerfiles for Linux containers, and will not work on Windows containers." Maybe you can elaborate on what "will not work" looks like.

**Comment by thaJeztah** [comment]
ping @estesp 😇

**Comment by estesp** [comment]
@mstanleyjones @thaJeztah updated; PTAL.

I did not combine the sections yet; I'm definitely willing to do so, but currently the anchor points are to each possible command in a `Dockerfile`; so I don't know what to name that anchor. Is it "ADD, COPY" or "ADD and COPY" or "ADD/COPY" ? :)

Given they are still distinct I added the same exact note about lack of capability on Windows to use the `--chown` feature.

File: docs/reference/builder.md [hunk 8]
```diff
@@ -938,15 +964,23 @@ guide](https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practi
 
 COPY has two forms:
 
-- `COPY <src>... <dest>`
-- `COPY ["<src>",... "<dest>"]` (this form is required for paths containing
+- `COPY [--chown=<user>:<group>] <src>... <dest>`
+- `COPY [--chown=<user>:<group>] ["<src>",... "<dest>"]` (this form is required for paths containing
```
Comment (thaJeztah) [inline docs/reference/builder.md]: lol, this will be a bit confusing as the first square brackets are to indicate "it's optional" and the second are literal brackets for JSON.

Don't have a better suggestion at this point though (other than having an example perhaps) - I'm okay with adding that in a follow up

File: docs/reference/builder.md (hunk 4 above)
Comment (johnstep) [reply docs/reference/builder.md]: @estesp What I've seen is that passing `--chown` succeeds on Windows, if a _number_ is supplied, but fails for a name since the code attempts to resolve the name by looking in `\etc\passwd` on Windows, which does not make sense.

Failing outright probably makes more sense, or at least warning, but I would prefer a more deliberate check and message. We have an issue on adding a warning, but it needs some more details: https://github.com/moby/moby/issues/35026

@thaJeztah We'll look into how the owner is set on Windows. From a quick test, it seems to be set differently based on which directory a file is copied to, but we need to verify.

**Comment by estesp** [comment]
This feature went into the 17.09 release, so it currently undocumented as far as online docs. Is there anything else not addressed here? Just don't want it to get lost for too long!

**Comment by karser** [comment]
@estesp keep track changelog on docker-ce [releases](https://github.com/docker/docker-ce/releases) page to stay current

**Comment by mstanleyjones** [comment]
@estesp well, this PR would have needed to be merged before `docker/cli` was merged into the appropriate `docker/docker-ce` branch.  The docs currently pull the `builder.md` from https://github.com/docker/docker-ce/blob/17.09/components/cli/docs/reference/builder.md.

**Comment by rcjsuen** [comment]
Unfortunately, this didn't make 17.10 either as far as I can tell. That's too bad but hey, it happens.

What outstanding issues are there left that are preventing it from being merged? I would like to add support for this flag in my project once the documentation has been finalized and is made available online.

**Comment by thaJeztah** [comment]
I'll go ahead and merge :+1: thanks!
